### PR TITLE
feat(manager/cargo): support git dependencies

### LINKED ISSUE #26531: feat(manager/cargo): support for git dependencies
### Discussed in https://github.com/renovatebot/renovate/discussions/26521

<div type='discussions-op-text'>

<sup>Originally posted by **richvdh** January  5, 2024</sup>
### Tell us more.

https://github.com/richvdh/renovate-cargo-git-test is an example of a Cargo project with a git dependency. [`Cargo.lock`](https://github.com/richvdh/renovate-cargo-git-test/blob/main/Cargo.lock#L8) is out of date, and can be updated with `cargo update -p empty-git-lib`.

Renovate reports:

```json
{
  "baseBranch": "main"
  "config": {
    "cargo": [
      {
        "deps": [
          {
            "currentValue": "0.1.0",
            "datasource": "crate",
            "depName": "empty-git-lib",
            "depType": "dependencies",
            "lockedVersion": "0.1.0",
            "managerData": {
              "nestedVersion": true
            },
            "packageName": "empty-git-lib",
            "skipReason": "git-dependency",
            "updates": []
          }
        ],
        "lockFiles": [
          "Cargo.lock"
        ],
        "packageFile": "Cargo.toml"
      }
    ]
  }
}
```

My understanding, looking at the [source](https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/cargo/extract.ts#L49-L92), is that Renovate will refuse to touch any dependency that comes from a git repo. It would be great if support could be added!</div>

Reproduction: https://github.com/renovate-reproductions/26531

### PR DESCRIPTION
**Pull request by mkniewallner**
## Changes

Add support for git dependencies in `cargo` manager.

## Context

Closes #26531.

Cargo specification: https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-git-repositories.

There is [one change](https://github.com/renovatebot/renovate/pull/32235#discussion_r1826587907) I am really not sure about.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

Tested against https://github.com/mkniewallner/renovate-rust-git-dependencies.

### CODE REVIEW

File: lib/modules/manager/cargo/artifacts.ts [hunk 1]
```diff
@@ -130,13 +131,23 @@ async function updateArtifactsImpl(
...
+      const nonCrateDep = updatedDeps.find(
+        (dep) => dep.datasource !== CrateDatasource.id,
+      );
+      const crateDepWithoutLockedVersion = updatedDeps.find(
+        (dep) => !dep.lockedVersion && dep.datasource === CrateDatasource.id,
+      );
+      // Non-crate dependencies (like git ones) do not have locked versions.
+      // For crate dependencies, not having a locked version is not expected.
+      // In both situations, perform a regular workspace lockfile update.
+      if (nonCrateDep || crateDepWithoutLockedVersion) {
```
Comment (mkniewallner) [inline lib/modules/manager/cargo/artifacts.ts:143]: Git dependencies do not have locked versions in the lock file, so without this, while we would correctly call `cargoUpdate` instead of `cargoUpdatePrecise`, without this change, we would display a warning that appears in the dependency dashboard when updating git dependencies.

Not really a huge fan of the overall logic, but I did not really find cleaner ways to achieve this.

Comment (rarkins) [reply lib/modules/manager/cargo/artifacts.ts:143]: Could you split this back to Discussion - either existing or new - assuming that a design decision needs making? It's hard to follow threads in PR reviews

**Comment by okkero** [comment]
What's keeping this right now? Can I help?

**Comment by okkero** [comment]
How can I contribute to this? @rarkins @mkniewallner

**Comment by rarkins** [comment]
@okkero the best way would be for you to sync this branch into your own fork, then fix conflicts + coverage

**Comment by okkero** [comment]
@rarkins Thanks. I might be able to take a stab at it later this week.

**Comment by okkero** [comment]
I've made a fork of this PR where I've resolved merge conflicts: https://github.com/renovatebot/renovate/pull/36289
